### ISSUE TITLE
Version matching for prereleases contradicts PEP 440

### ISSUE THREAD

**Issue by konstin** [pypa/packaging#617]
PEP 440 [says](https://peps.python.org/pep-0440/#version-matching):

> For purposes of prefix matching, the pre-release segment is considered to have an implied preceding `.`, so given the version `1.1a1`, the following clauses would match or not as shown:
>
>     == 1.1        # Not equal, so 1.1a1 does not match clause
>     == 1.1a1      # Equal, so 1.1a1 matches clause
>     == 1.1.*      # Same prefix, so 1.1a1 matches clause

However, as of 606c71a, the implementation contradicts this:

```python
>>> from packaging.specifiers import SpecifierSet
>>> from packaging.version import Version
>>> Version("1.1a1") in SpecifierSet("== 1.1.*")
False
```

**Comment by uranusjr** [comment]
https://peps.python.org/pep-0440/#handling-of-pre-releases

> Pre-releases of any kind, including developmental releases, are implicitly excluded from all version specifiers, unless they are already present on the system, explicitly requested by the user, or if the only available version that satisfies the version specifier is a pre-release.

It’s possible to make a case this is an example of the user explicitly requesting a prerelease, but PEP 440 is quite unclear about this.

```pycon
>>> SpecifierSet("== 1.1.*").contains(Version("1.1a1"), prereleases=True)
True
```

**Comment by konstin** [comment]
yeah PEP 440 seems to be rather fuzzy about prereleases and local versions in general. So from a `pypa/packaging` perspective, the examples in PEP 440 are to be seen as `SpecifierSet(_, prereleases=True)`?

**Comment by uranusjr** [comment]
Yes. Currently “requested by the user” can be achieved by either passing `prereleases` explicitly, or use a prerelease version somewhere in the specifier (e.g. `>= 1.1.a0`). I think I proposed making the `prereleases` a required flag instead a while ago (since even experienced contributors get tripped up by this occasionally), but it’s not easy to get everyone migrated.

**Comment by pradyunsg** [comment]
This nuance is covered in the examples at https://packaging.pypa.io/en/latest/specifiers.html#packaging.specifiers.Specifier.__contains__ and https://packaging.pypa.io/en/latest/specifiers.html#packaging.specifiers.Specifier.contains

I really think PEP 440 being treated as a user-facing document within the Python community is a symptom of a problem but I digress. Is there any actionable suggestions here, for something we could do differently in the documentation or elsehwere to help make this clearer?

**Comment by konstin** [comment]
> Yes. Currently “requested by the user” can be achieved by either passing `prereleases` explicitly, or use a prerelease version somewhere in the specifier (e.g. `>= 1.1.a0`). 

Thanks for clarifying

> I think I proposed making the `prereleases` a required flag instead a while ago (since even experienced contributors get tripped up by this occasionally), but it’s not easy to get everyone migrated.

Currently, the parameter allows for some strange behavior:

```python
from packaging.version import Version
from packaging.specifiers import SpecifierSet

assert Version("1.1a1") in SpecifierSet("==1.1a1")
assert Version("1.1a1") not in SpecifierSet("==1.1a1", prereleases=False)
```

again, this is "[...], explicitly requested by the user, or if the only available version that satisfies the version specifier is a pre-release." being fuzzy enough that's it's unclear what the correct behavior is

> This nuance is covered in the examples at https://packaging.pypa.io/en/latest/specifiers.html#packaging.specifiers.Specifier.__contains__ and https://packaging.pypa.io/en/latest/specifiers.html#packaging.specifiers.Specifier.contains. 

Sorry, i should have read the documentation properly.

> Are there any actionable suggestions here, for something we could do differently in the documentation or elsehwere to help make this clearer?

I think it would be better if the examples in PEP 440 explicitly stated that difference. I gave it a try at https://github.com/python/peps/pull/2898

> FWIW, I really think PEP 440 being treated as a user-facing document within the Python community is a symptom of a problem but I digress.

Same and same.

Would it be possible to copy over PEP 440 verbatim to the current stub page and then start editing that outside of the PEP constraints, e.g. adding the version sorting test from this repository as an example?

**Comment by brettcannon** [comment]
All packaging PEPs should be covered in https://packaging.python.org/en/latest/specifications/ , and so it's probably best to just port PEP 440 over there and then try to provide better examples after it gets merged.

But as Pradyun asked, is there something specific here for `packaging` you are suggesting we change?

**Comment by J3ronimo** [comment]
In extension,

```python
SpecifierSet(">= 1.0.0").contains("1.0.0rc1", prereleases=True)
```

returns `False`.

It makes sense that the `>=` operator considers a prerelease lower than the release of the same version, since the prerelesae comes earlier in the timeline. But since prereleases were explicitly requested, I would have expected the this to only look at the "1.0.0" prefix, same as with the "1.1." prefix in the OP example above.

That's what the comment in the third line suggests:

```
== 1.1        # Not equal, so 1.1a1 does not match clause
== 1.1a1      # Equal, so 1.1a1 matches clause
== 1.1.*      # Same prefix, so 1.1a1 matches clause if pre-releases are requested
